docs/docs/development/events.md

```diff
@@ -0,0 +1,178 @@
+---
+title: ZMK Events
+sidebar_label: ZMK Events
+---
+
+ZMK makes use of events to decouple individual components such as behaviors and peripherals from the core functionality. For this purpose, ZMK has implemented its own event manager. This page is a (brief) overview of the functionality and methods exposed by the event manager. To see what events exist and what data they contain, it is best to view the corresponding [event header files](https://github.com/zmkfirmware/zmk/tree/main/app/include/zmk/events) directly.
+
+Including `zmk/event_manager.h` is required for any interaction with the event system.
+
+## Generic Events
+
+The generic event type is `struct zmk_event_t`. This struct looks like this:
+
+```c
+typedef struct {
+    const struct zmk_event_type *event;
+    uint8_t last_listener_index;
+} zmk_event_t;
+```
+
+In memory, a raised event **always** consists of a `zmk_event_t` struct **followed immediately afterwards** by the struct containing the data for the actual event. The contents of `const struct zmk_event_type *event` allows us to identify which event a particular event actually is, so that we may safely access its data. The type of the combination of a `zmk_event_t` struct with a specific event data struct, e.g. `struct zmk_specific_thing_happened`, is `struct zmk_specific_thing_happened_event`:
+
+```c
+struct zmk_specific_thing_happened_event {
+    zmk_event_t header;
+    struct zmk_specific_thing_happened data;
+};
+```
+
+To obtain the underlying data from a generic event, the following function is used:
+
+```c
+struct zmk_specific_thing_happened *as_specific_thing_happened(const zmk_event_t *eh);
+```
+
+This method takes in a pointer to a `zmk_event_t` (which is actually a pointer to a specific event, such as `zmk_specific_thing_happened_event`), and will return the underlying `zmk_specific_thing_happened` data struct if the `zmk_event_t` header indicates that the generic event pointer is indeed a pointer to a `zmk_specific_thing_happened_event`. If the type of the event does not match the function, then the function will return `NULL`. By convention, `zmk_event_t` pointer arguments are named `eh`, short for "event header".
```

Comment:
This is perhaps outside the scope of the docs but it is confusing me as I read this so I'll ask: Why is the arg type `zmk_event_t` here? It seems like this is not correct, as it is actually a specific event type, like you mentioned here.